app/src/main/java/at/bitfire/icsdroid/SyncWorker.kt

```diff
@@ -70,29 +102,94 @@ class SyncWorker(
+            // Get a list of all the subscriptions from the database
+            val subscriptions = subscriptionsDao.getAll().toMutableList()
+
+            // Get a provider from the application context, or return failure
+            val provider = DatabaseAndroidInterface.getProvider(applicationContext) ?: return Result.failure()
+            // If there's a provider available, get all the calendars available in the system
+            val calendars = AndroidCalendar.find(
+                account,
+                provider,
+                LocalCalendar.Factory,
+                null,
+                null,
+            )
+
+            // Check that all the calendars have a matching subscription
+            for (calendar in calendars) {
+                val id = calendar.id
+                val match = subscriptions.find { subscription -> subscription.id == id }
+                // If there's already a calendar matching the subscription, continue
+                if (match != null) continue
+                try {
+                    // Otherwise, create a subscription for the calendar
+                    val newSubscription = Subscription.fromCalendar(calendar)
+                    // Add it to the database
+                    subscriptionsDao.add(newSubscription)
+                    // And add it to `subscriptions` so it gets processed now.
+                    subscriptions.add(newSubscription)
+                    Log.i(TAG, "The calendar #${calendar.id} didn't have a matching subscription. Just created it.")
+                } catch (e: Exception) {
+                    Log.e(TAG, "Could not create subscription from calendar. Migration failed.", e)
+                    continue
+                }
+            }
+
+            // Migrate all credentials to the database
+            val credentialsDao = database.credentialsDao()
+            @Suppress("DEPRECATION") val oldCredentials = CalendarCredentials(applicationContext)
+            for (subscription in subscriptions) {
+                val databaseAndroidInterface = DatabaseAndroidInterface(applicationContext, subscription)
+                val calendar = databaseAndroidInterface.getCalendar()
+                calendar
+                    // Get the credentials that might be stored for the subscription
+                    .let { oldCredentials.get(it) }
+                    // Take only if there's an username and password
+                    .takeIf { (u, p) -> u != null && p != null }
+                    // Convert the username and password into Credential
+                    ?.let { Credential(subscription.id, it.first!!, it.second!!) }
+                    // Store the credential in the database
+                    ?.let { credentialsDao.put(it.subscriptionId, it.username, it.password) }
+                    // Remove the credential from shared preferences
+                    ?.also { oldCredentials.put(calendar, null, null) }
+            }
```

Comment:
We should put migration code to a separate method, so that the syncWorker looks like that:

```
// handle calendars that don't have a Subscription
migrateLegacyCalendars()

// sync subscriptions to calendars:
// - create/update/delete calendars according to the Subscriptions
// - sync each subscription/calendar
syncCalendars()
```

The `migrateLegacyCalendars()` function should only be responsible for migrating and be as separated as possible.